### ISSUE TITLE
Language markup (`lang="en") gets injected into strings with `rdf:HTML` datatype

### ISSUE THREAD

**Issue by BigBlueHat** [RDFLib/pyrdfa3#25]
Given this (rather silly) RDFa markup:
```html
<section lang="de" resource="#">
<h2>Keywords</h2>
<ul>
  <li property="schema:keywords">Gift</li>
  <li property="schema:keywords" lang="en">Gift</li>
  <li property="schema:keywords" datatype="rdf:HTML">CO<sub>2</sub> Gift</li>
</ul>
</section>
```
results in:
```turtle
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix schema: <http://schema.org/> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<#> schema:keywords "CO<sub lang=\"de\">2</sub> Gift"^^rdf:HTML,
        "Gift"@de,
        "Gift"@en .
```

The `lang="de"` (in this case) is injected into any outer-most markup. In the scenario above that generates some nonsensical markup--which the author didn't intend. 😃 

I don't believe this is intended behavior (the [Ruby Distiller](http://rdf.greggkellogg.net/distiller?command=serialize&format=rdfa&output_format=turtle) lacks this bug), so thought I'd report it.

Cheers!
🎩

**Comment by BigBlueHat** [comment]
Correct. I did mean `datatype="rdf:HTML"` (I'd also been trying to see how `rdf:XML` was handled for comparison...and copied from the wrong window 😛).

Does the RDFa Core spec specify how the HTML content should/must be altered to inject the language? I couldn't find it, and the Ruby Distiller, EasyPHP, and GreenTurtle (at least) all lacked this behavior.

Context:
I'm working with some content model colleagues at Wiley to narrow in on expected/specified behavior around language encodings and extractions from (and into) RDFa.

**Comment by BigBlueHat** [comment]
Actually...I'm realizing that first post uses all the wrong examples... 🙄 I'll update it to reflect what *actually* breaks and is confusing.

**Comment by iherman** [comment]
Actually, no, because there were some timing issues. In [HTML+RDFa 1.1](https://www.w3.org/TR/2015/REC-html-rdfa-20150317/) the status says:

> The specification makes use of the rdf:HTML datatype. This feature is non-normative, because the equality of the literal values depend on DOM4 [dom4], a specification that has not yet reached W3C Recommendation status. 

 It also says in the document:

>  the value of the HTML Literal is a string created by serializing all child nodes to text. This applies to all nodes that are descendants of the current element, not including the element itself. The HTML Literal is given a datatype of http://www.w3.org/1999/02/22-rdf-syntax-ns#HTML as defined in Section 5.2: The rdf:HTML Datatype of [rdf11-concepts]. 

That is all there is.

It is unclear to me  what is good or bad; this language aspect has never been properly defined (bug in the spec?). I would argue that, by virtue of having set the language to German explicitly, the values in the third `<li>` are to be taken as German. That is reproduced in the `rdf:HTML` content which does reflect the language setting which would be otherwise lost. But, as I said, this is not properly specified, in my view, in the RDFa spec.

**Comment by BigBlueHat** [comment]
(intro examples actually correct now 😏)

What happens currently is that the language gets injected into any available markup, but no surrounding markup is introduced (thankfully). This currently happens for *both* `rdf:HTML` and `rdf:XMLLiteral` `datatype`'d content.

The problem (as you can see in the example) is that "guessing" about the markup's language may create things like `<sub lang="de">2</sub>` (which is nonsensical), and also creates varying structures (which breaks equivalence) across multiple parsing of the same content--i.e. you'll get different object values depending on which parser you use.

DOM parsers (from what I've seen anyhow) don't "persist" a parent's language on to the child, so if the child is extracted ("serializing all child nodes to text") then the only language information that would be present (using the DOM) is whatever is present on each of those child nodes--i.e. an ancestor node's language only applies to children when they are currently children.

Consequently, this seems like a bug. 🐛 🐞 🐝 😄

**Comment by iherman** [comment]
@BigBlueHat, I accept it as a bug. I have just pushed a commit that should solve this issue.

However, while we are at it, there is a more general question on what version to really use...

For all kinds of awkwardnesses (mostly related to the way the original RDFa service was deployed on W3C when git was still not so widely used), there are two implementations of RDFa+RDFLib around.

1. This repo, which is a layer _on top_ of RDFLib. (This was my original development that led to the [RDFa service at W3C](https://www.w3.org/2012/pyRdfa/).)
2. There is a version of this code that has been incorporated into the core RDFLib library as one of the parsers. The core code is the same, except that the input and output is organized differently.

Without getting into history, the current situation is a bit messy. As of today, I'd consider (1) as being more or less stalled, because having both libraries around is obviously messy and, I believe, an incorporated parser is better for users. Hence it would be better if you used the parser of the core RDFLib.

Except that... there is a [PR](https://github.com/RDFLib/rdflib/pull/828), prepared by @gromgull, that proposes to _remove_ that parser from the core RDFLib library and rely on this repo content as an plugin (except that, I presume, that would require some more development to make it usable as a plugin). As I said, personally, I would prefer to freeze this rep, and consider the current version of RDFLib (ie, with the parser approach) as being the authoritative one but, to be fair, the work on RDFLib is really managed by @gromgull; having moved out of this area altogether, I am not sure my opinion counts. Maybe he can chime in at this point.

In any case, I have handled the bug in the RDFLib parser version as well as well: I have pushed a separate branch [`rdfa-html-datatype-lang-error`](https://github.com/RDFLib/rdflib/tree/rdfa-html-datatype-lang-error), branched off from the latest `master` and that can be merged into master to get this specific bug taken care off in that repo, too. I let @gromgull tell me if I should do a PR.

Yes, a bit messy, and I am partially to blame. But that is where we are...
